File: onecycle_lr.py
import tensorflow as tf
from tensorflow.keras.callbacks import Callback
import math
import numpy as np

# Import the global variables
from globals import log_queue


class OneCycleLR(Callback):
    """
    Enhanced Implementation of the One Cycle Learning Rate Policy.
    
    The 1cycle policy anneals the learning rate from an initial learning rate to a
    maximum learning rate and then back to an even lower final learning rate. This policy was
    proposed by Leslie Smith in the paper 'Super-Convergence: Very Fast Training of 
    Neural Networks Using Large Learning Rates'.
    
    This implementation adds:
    - Better cosine annealing with more controlled transitions
    - Weight decay decoupled from learning rate
    - Improved monitoring and logging
    
    Args:
        max_lr: Maximum learning rate.
        steps_per_epoch: Number of steps per epoch.
        epochs: Total number of epochs.
        min_lr: Minimum learning rate (starting). Default is max_lr/10.
        final_div_factor: Final LR will be max_lr/final_div_factor. Default is 100.
        warmup_pct: Percentage of total iterations for warmup phase. Default is 0.3.
        current_lr_var: Tkinter variable to update UI with current learning rate.
        momentum_range: Tuple of (max_momentum, min_momentum) to vary momentum inversely to lr.
        weight_decay: Weight decay factor. Set to None to disable. Default is None.
        verbose: Whether to print verbose output. Default is 1.
    """
    def __init__(self, max_lr, steps_per_epoch, epochs, min_lr=None, 
                 final_div_factor=100, warmup_pct=0.3, current_lr_var=None, 
                 momentum_range=(0.95, 0.85), weight_decay=None, verbose=1):
        super(OneCycleLR, self).__init__()
        self.max_lr = max_lr
        self.min_lr = min_lr if min_lr is not None else max_lr / 10
        self.final_lr = max_lr / final_div_factor  # End with even lower LR
        self.steps_per_epoch = steps_per_epoch
        self.total_steps = steps_per_epoch * epochs
        self.warmup_steps = int(self.total_steps * warmup_pct)
        self.annealing_steps = self.total_steps - self.warmup_steps
        self.current_step = 0
        self.current_lr_var = current_lr_var
        self.history = {'lr': [], 'momentum': [], 'weight_decay': []}
        self.initialized = False
        self.verbose = verbose
        
        # Momentum scheduling (inverse to learning rate)
        self.max_momentum, self.min_momentum = momentum_range
        
        # Weight decay scheduling
        self.weight_decay = weight_decay
        self.weight_decay_schedule = []
        if weight_decay is not None:
            # Create weight decay schedule - follow LR pattern but scaled
            self.initial_weight_decay = weight_decay
            self.max_weight_decay = weight_decay * 10  # Increase during warmup
            self.final_weight_decay = weight_decay / 10  # Lower at the end
        
    def _calculate_lr(self, step):
        """Calculate learning rate for current step according to schedule"""
        if step < self.warmup_steps:
            # Linear warmup phase
            progress = step / self.warmup_steps
            return self.min_lr + (self.max_lr - self.min_lr) * progress
        else:
            # Cosine annealing phase
            progress = (step - self.warmup_steps) / self.annealing_steps
            # Smoother cosine annealing with a slight offset to avoid sudden changes
            cosine_decay = 0.5 * (1 + np.cos(np.pi * progress))
            return self.final_lr + (self.max_lr - self.final_lr) * cosine_decay
    
    def _calculate_momentum(self, step):
        """Calculate momentum for current step - inverse to learning rate"""
        if step < self.warmup_steps:
            # Linear decrease during warmup
            progress = step / self.warmup_steps
            return self.max_momentum - (self.max_momentum - self.min_momentum) * progress
        else:
            # Linear increase during annealing
            progress = (step - self.warmup_steps) / self.annealing_steps
            cosine_decay = 0.5 * (1 + np.cos(np.pi * progress))
            return self.min_momentum + (self.max_momentum - self.min_momentum) * cosine_decay
    
    def _calculate_weight_decay(self, step):
        """Calculate weight decay for current step if enabled"""
        if self.weight_decay is None:
            return None
            
        if step < self.warmup_steps:
            # Linear increase during warmup
            progress = step / self.warmup_steps
            return self.initial_weight_decay + (self.max_weight_decay - self.initial_weight_decay) * progress
        else:
            # Cosine annealing during cooldown
            progress = (step - self.warmup_steps) / self.annealing_steps
            cosine_decay = 0.5 * (1 + np.cos(np.pi * progress))
            return self.final_weight_decay + (self.max_weight_decay - self.final_weight_decay) * cosine_decay
        
    def on_train_begin(self, logs=None):
        # Initialize with min_lr
        if not hasattr(self.model, 'optimizer') or self.model.optimizer is None:
            log_queue.put("Warning: Optimizer not found in model. Cannot set learning rate.")
            return
            
        # Set initial learning rate
        tf.keras.backend.set_value(self.model.optimizer.lr, self.min_lr)
        
        # Initialize momentum
        if hasattr(self.model.optimizer, 'momentum'):
            tf.keras.backend.set_value(self.model.optimizer.momentum, self.max_momentum)
            
        # Setup weight decay for AdamW
        if self.weight_decay is not None:
            if hasattr(self.model.optimizer, 'decay'):
                self.supports_weight_decay = True
                tf.keras.backend.set_value(self.model.optimizer.decay, self.initial_weight_decay)
            elif hasattr(self.model.optimizer, 'weight_decay'):
                self.supports_weight_decay = True
                tf.keras.backend.set_value(self.model.optimizer.weight_decay, self.initial_weight_decay)
            else:
                self.supports_weight_decay = False
                if self.verbose > 0:
                    log_queue.put("Warning: Weight decay requested but optimizer doesn't support it.")
            
        # Update UI
        if self.current_lr_var:
            try:
                self.current_lr_var.set(f"Current LR: {self.min_lr:.6f}")
            except Exception as e:
                log_queue.put(f"Error updating LR display: {str(e)}")
                
        # Only log this information once on the first training run
        if not self.initialized and self.verbose > 0:
            log_queue.put(f"OneCycleLR: Warmup: {self.warmup_steps} steps, Annealing: {self.annealing_steps} steps")
            log_queue.put(f"OneCycleLR: LR range: [{self.min_lr:.6f}, {self.max_lr:.6f}, {self.final_lr:.6f}]")
            log_queue.put(f"OneCycleLR: Momentum range: [{self.min_momentum:.4f}, {self.max_momentum:.4f}]")
            if self.weight_decay is not None and self.supports_weight_decay:
                log_queue.put(f"OneCycleLR: Weight decay active: initial={self.initial_weight_decay}")
            self.initialized = True
    
    def on_epoch_begin(self, epoch, logs=None):
        # Log current settings at the beginning of each epoch
        if self.verbose > 0:
            current_lr = float(tf.keras.backend.get_value(self.model.optimizer.lr))
            
            momentum_str = ""
            if hasattr(self.model.optimizer, 'momentum'):
                current_momentum = float(tf.keras.backend.get_value(self.model.optimizer.momentum))
                momentum_str = f", momentum={current_momentum:.4f}"
                
            weight_decay_str = ""
            if self.weight_decay is not None and self.supports_weight_decay:
                if hasattr(self.model.optimizer, 'decay'):
                    current_wd = float(tf.keras.backend.get_value(self.model.optimizer.decay))
                else:
                    current_wd = float(tf.keras.backend.get_value(self.model.optimizer.weight_decay))
                weight_decay_str = f", weight_decay={current_wd:.6f}"
            
            log_queue.put(f"Epoch {epoch+1} starting with lr={current_lr:.6f}{momentum_str}{weight_decay_str}")
                
    def on_batch_begin(self, batch, logs=None):
        # Skip if we've gone beyond total steps
        if self.current_step >= self.total_steps:
            return
            
        # Calculate the current learning rate, momentum, and weight decay
        lr = self._calculate_lr(self.current_step)
        
        # Set the learning rate
        tf.keras.backend.set_value(self.model.optimizer.lr, lr)
        
        # Set momentum if available
        if hasattr(self.model.optimizer, 'momentum'):
            momentum = self._calculate_momentum(self.current_step)
            tf.keras.backend.set_value(self.model.optimizer.momentum, momentum)
            self.history['momentum'].append(momentum)
        
        # Set weight decay if supported
        if self.weight_decay is not None and self.supports_weight_decay:
            weight_decay = self._calculate_weight_decay(self.current_step)
            if hasattr(self.model.optimizer, 'decay'):
                tf.keras.backend.set_value(self.model.optimizer.decay, weight_decay)
            else:
                tf.keras.backend.set_value(self.model.optimizer.weight_decay, weight_decay)
            self.history['weight_decay'].append(weight_decay)
            
        # Update UI variable if provided (every 5 batches to reduce overhead)
        if self.current_lr_var and (batch % 5 == 0 or batch == 0):
            try:
                self.current_lr_var.set(f"Current LR: {lr:.6f}")
            except Exception as e:
                log_queue.put(f"Error updating LR display: {str(e)}")
        
        # Store history for plotting
        self.history['lr'].append(lr)
        
        # Increment step counter
        self.current_step += 1
        
    def on_epoch_end(self, epoch, logs=None):
        if self.verbose > 0:
            # Log current learning rate at the end of each epoch
            current_lr = float(tf.keras.backend.get_value(self.model.optimizer.lr))
            
            # Create momentum message if applicable
            momentum_msg = ""
            if hasattr(self.model.optimizer, 'momentum'):
                current_momentum = float(tf.keras.backend.get_value(self.model.optimizer.momentum))
                momentum_msg = f" with momentum: {current_momentum:.4f}"
                
            # Create weight decay message if applicable
            weight_decay_msg = ""
            if self.weight_decay is not None and self.supports_weight_decay:
                if hasattr(self.model.optimizer, 'decay'):
                    current_wd = float(tf.keras.backend.get_value(self.model.optimizer.decay))
                else: 
                    current_wd = float(tf.keras.backend.get_value(self.model.optimizer.weight_decay))
                weight_decay_msg = f", weight_decay: {current_wd:.6f}"
                
            log_queue.put(f"Epoch {epoch+1} completed with lr: {current_lr:.6f}{momentum_msg}{weight_decay_msg}")
        
        # Update UI
        if self.current_lr_var:
            try:
                current_lr = float(tf.keras.backend.get_value(self.model.optimizer.lr))
                self.current_lr_var.set(f"Current LR: {current_lr:.6f}")
            except Exception as e:
                log_queue.put(f"Error updating LR display: {str(e)}")
                
    def on_train_end(self, logs=None):
        if self.verbose > 0:
            log_queue.put("OneCycleLR schedule completed")
            
    def get_lr_schedule(self):
        """Returns the full learning rate schedule as a numpy array for plotting"""
        return np.array([self._calculate_lr(i) for i in range(self.total_steps)])
        
    def get_momentum_schedule(self):
        """Returns the full momentum schedule as a numpy array for plotting"""
        return np.array([self._calculate_momentum(i) for i in range(self.total_steps)])
        
    def get_weight_decay_schedule(self):
        """Returns the full weight decay schedule as a numpy array for plotting"""
        if self.weight_decay is None:
            return None
        return np.array([self._calculate_weight_decay(i) for i in range(self.total_steps)])


class SWA(Callback):
    """
    Stochastic Weight Averaging callback.
    
    Implements the Stochastic Weight Averaging (SWA) technique for improving
    generalization in deep learning models by averaging weights from multiple
    points in the training trajectory.
    
    Args:
        start_epoch: The epoch to start averaging from. Default is 10.
        swa_freq: Frequency of weight averaging in epochs. Default is 5.
        verbose: Verbosity level. Default is 1.
    """
    def __init__(self, start_epoch=10, swa_freq=5, verbose=1):
        super(SWA, self).__init__()
        self.start_epoch = start_epoch
        self.swa_freq = swa_freq
        self.verbose = verbose
        self.swa_weights = None
        self.swa_count = 0
        
    def on_epoch_end(self, epoch, logs=None):
        # Check if we should update SWA
        if epoch + 1 >= self.start_epoch and (epoch + 1 - self.start_epoch) % self.swa_freq == 0:
            # Get current model weights
            current_weights = self.model.get_weights()
            
            # Update SWA weights
            if self.swa_weights is None:
                # First time - just copy the weights
                self.swa_weights = current_weights
            else:
                # Update running average
                for i, w in enumerate(current_weights):
                    self.swa_weights[i] = (self.swa_weights[i] * self.swa_count + w) / (self.swa_count + 1)
            
            # Update count
            self.swa_count += 1
            
            if self.verbose > 0:
                log_queue.put(f"SWA: Updated weights average at epoch {epoch+1} (total models: {self.swa_count})")
    
    def on_train_end(self, logs=None):
        # Apply SWA weights when training ends
        if self.swa_weights is not None:
            if self.verbose > 0:
                log_queue.put(f"SWA: Applying averaged weights from {self.swa_count} models")
            
            # Store original weights to allow recovery
            self.original_weights = self.model.get_weights()
            
            # Set SWA weights to model
            self.model.set_weights(self.swa_weights)
            
            if self.verbose > 0:
                log_queue.put("SWA: Weights successfully applied")
        else:
            if self.verbose > 0:
                log_queue.put("SWA: No weights were averaged, model remains unchanged")
                
    def get_original_weights(self):
        """Return the original (non-SWA) weights if available"""
        if hasattr(self, 'original_weights'):
            return self.original_weights
        return None
        
    def reset_original_weights(self):
        """Reset model to the original weights before SWA"""
        if hasattr(self, 'original_weights'):
            self.model.set_weights(self.original_weights)
            return True
        return False

File: tfjs/onecycle_lr.js
// onecycle_lr.js
// Implementation of the One Cycle Learning Rate policy for TensorFlow.js

import { logQueue } from './globals.js';

// OneCycleLR class for learning rate scheduling
class OneCycleLR {
  /**
   * Enhanced Implementation of the One Cycle Learning Rate Policy.
   * 
   * @param {Object} config Configuration object
   * @param {number} config.maxLr Maximum learning rate
   * @param {number} config.stepsPerEpoch Number of steps per epoch
   * @param {number} config.epochs Total number of epochs
   * @param {number} config.minLr Minimum learning rate (starting). Default is maxLr/10
   * @param {number} config.finalDivFactor Final LR will be maxLr/finalDivFactor. Default is 100
   * @param {number} config.warmupPct Percentage of total iterations for warmup phase. Default is 0.3
   * @param {Function} config.onLrUpdate Function to call when LR updates (for UI updates)
   * @param {Array} config.momentumRange Tuple of (max_momentum, min_momentum) to vary momentum inversely to lr
   * @param {number} config.weightDecay Weight decay factor. Set to null to disable
   * @param {number} config.verbose Whether to print verbose output
   */
  constructor({
    maxLr,
    stepsPerEpoch,
    epochs,
    minLr = null,
    finalDivFactor = 100,
    warmupPct = 0.3,
    onLrUpdate = null,
    momentumRange = [0.95, 0.85],
    weightDecay = null,
    verbose = 1
  }) {
    this.maxLr = maxLr;
    this.minLr = minLr !== null ? minLr : maxLr / 10;
    this.finalLr = maxLr / finalDivFactor;
    this.stepsPerEpoch = stepsPerEpoch;
    this.totalSteps = stepsPerEpoch * epochs;
    this.warmupSteps = Math.floor(this.totalSteps * warmupPct);
    this.annealingSteps = this.totalSteps - this.warmupSteps;
    this.currentStep = 0;
    this.onLrUpdate = onLrUpdate;
    this.history = {
      lr: [],
      momentum: [],
      weightDecay: []
    };
    this.initialized = false;
    this.verbose = verbose;
    
    // Momentum scheduling (inverse to learning rate)
    [this.maxMomentum, this.minMomentum] = momentumRange;
    
    // Weight decay scheduling
    this.weightDecay = weightDecay;
    this.weightDecaySchedule = [];
    
    if (weightDecay !== null) {
      // Create weight decay schedule - follow LR pattern but scaled
      this.initialWeightDecay = weightDecay;
      this.maxWeightDecay = weightDecay * 10;  // Increase during warmup
      this.finalWeightDecay = weightDecay / 10;  // Lower at the end
    }
  }
  
  /**
   * Calculate learning rate for current step according to schedule
   * @param {number} step Current step
   * @return {number} Learning rate for this step
   */
  _calculateLr(step) {
    if (step < this.warmupSteps) {
      // Linear warmup phase
      const progress = step / this.warmupSteps;
      return this.minLr + (this.maxLr - this.minLr) * progress;
    } else {
      // Cosine annealing phase
      const progress = (step - this.warmupSteps) / this.annealingSteps;
      // Smoother cosine annealing with a slight offset to avoid sudden changes
      const cosineDecay = 0.5 * (1 + Math.cos(Math.PI * progress));
      return this.finalLr + (this.maxLr - this.finalLr) * cosineDecay;
    }
  }
  
  /**
   * Calculate momentum for current step - inverse to learning rate
   * @param {number} step Current step
   * @return {number} Momentum for this step
   */
  _calculateMomentum(step) {
    if (step < this.warmupSteps) {
      // Linear decrease during warmup
      const progress = step / this.warmupSteps;
      return this.maxMomentum - (this.maxMomentum - this.minMomentum) * progress;
    } else {
      // Linear increase during annealing
      const progress = (step - this.warmupSteps) / this.annealingSteps;
      const cosineDecay = 0.5 * (1 + Math.cos(Math.PI * progress));
      return this.minMomentum + (this.maxMomentum - this.minMomentum) * cosineDecay;
    }
  }
  
  /**
   * Calculate weight decay for current step if enabled
   * @param {number} step Current step
   * @return {number|null} Weight decay for this step or null if disabled
   */
  _calculateWeightDecay(step) {
    if (this.weightDecay === null) {
      return null;
    }
    
    if (step < this.warmupSteps) {
      // Linear increase during warmup
      const progress = step / this.warmupSteps;
      return this.initialWeightDecay + (this.maxWeightDecay - this.initialWeightDecay) * progress;
    } else {
      // Cosine annealing during cooldown
      const progress = (step - this.warmupSteps) / this.annealingSteps;
      const cosineDecay = 0.5 * (1 + Math.cos(Math.PI * progress));
      return this.finalWeightDecay + (this.maxWeightDecay - this.finalWeightDecay) * cosineDecay;
    }
  }
  
  /**
   * Called at the start of training
   * @param {Object} model TensorFlow.js model
   */
  onTrainBegin(model) {
    // Set initial learning rate
    if (model && model.optimizer) {
      model.optimizer.learningRate = this.minLr;
      
      // Initialize momentum if supported
      if (model.optimizer.momentum !== undefined) {
        model.optimizer.momentum = this.maxMomentum;
      }
      
      // Setup weight decay for Adam (not directly supported in tfjs)
      // Will be handled in training loop
    }
    
    // Update UI
    if (this.onLrUpdate) {
      try {
        this.onLrUpdate(this.minLr);
      } catch (e) {
        console.error("Error updating LR display:", e);
      }
    }
    
    // Log schedule information
    if (!this.initialized && this.verbose > 0) {
      logQueue.put(`OneCycleLR: Warmup: ${this.warmupSteps} steps, Annealing: ${this.annealingSteps} steps`);
      logQueue.put(`OneCycleLR: LR range: [${this.minLr.toFixed(6)}, ${this.maxLr.toFixed(6)}, ${this.finalLr.toFixed(6)}]`);
      logQueue.put(`OneCycleLR: Momentum range: [${this.minMomentum.toFixed(4)}, ${this.maxMomentum.toFixed(4)}]`);
      if (this.weightDecay !== null) {
        logQueue.put(`OneCycleLR: Weight decay active: initial=${this.initialWeightDecay}`);
      }
      this.initialized = true;
    }
  }
  
  /**
   * Called at the beginning of each epoch
   * @param {number} epoch Current epoch (0-indexed)
   * @param {Object} model TensorFlow.js model
   */
  onEpochBegin(epoch, model) {
    // Log current settings at the beginning of each epoch
    if (this.verbose > 0 && model && model.optimizer) {
      const currentLr = model.optimizer.learningRate;
      
      let momentumStr = "";
      if (model.optimizer.momentum !== undefined) {
        const currentMomentum = model.optimizer.momentum;
        momentumStr = `, momentum=${currentMomentum.toFixed(4)}`;
      }
      
      let weightDecayStr = "";
      if (this.weightDecay !== null) {
        const currentWd = this._calculateWeightDecay(this.currentStep);
        weightDecayStr = `, weight_decay=${currentWd.toFixed(6)}`;
      }
      
      logQueue.put(`Epoch ${epoch+1} starting with lr=${currentLr.toFixed(6)}${momentumStr}${weightDecayStr}`);
    }
  }
  
  /**
   * Called at the beginning of each batch
   * @param {number} batch Current batch within the epoch
   * @param {Object} model TensorFlow.js model
   */
  onBatchBegin(batch, model) {
    // Skip if we've gone beyond total steps
    if (this.currentStep >= this.totalSteps) {
      return;
    }
    
    // Calculate the current learning rate
    const lr = this._calculateLr(this.currentStep);
    
    // Set the learning rate
    if (model && model.optimizer) {
      model.optimizer.learningRate = lr;
      
      // Set momentum if available
      if (model.optimizer.momentum !== undefined) {
        const momentum = this._calculateMomentum(this.currentStep);
        model.optimizer.momentum = momentum;
        this.history.momentum.push(momentum);
      }
      
      // Weight decay is handled separately in the optimizer step
    }
    
    // Update UI variable if provided (every 5 batches to reduce overhead)
    if (this.onLrUpdate && (batch % 5 === 0 || batch === 0)) {
      try {
        this.onLrUpdate(lr);
      } catch (e) {
        console.error("Error updating LR display:", e);
      }
    }
    
    // Store history for plotting
    this.history.lr.push(lr);
    
    // Increment step counter
    this.currentStep++;
    
    // Return the current lr and weight decay for use in the optimizer
    return {
      lr,
      weightDecay: this.weightDecay !== null ? this._calculateWeightDecay(this.currentStep - 1) : null
    };
  }
  
  /**
   * Called at the end of each epoch
   * @param {number} epoch Current epoch (0-indexed)
   * @param {Object} model TensorFlow.js model
   */
  onEpochEnd(epoch, model) {
    if (this.verbose > 0 && model && model.optimizer) {
      // Log current learning rate at the end of each epoch
      const currentLr = model.optimizer.learningRate;
      
      // Create momentum message if applicable
      let momentumMsg = "";
      if (model.optimizer.momentum !== undefined) {
        const currentMomentum = model.optimizer.momentum;
        momentumMsg = ` with momentum: ${currentMomentum.toFixed(4)}`;
      }
      
      // Create weight decay message if applicable
      let weightDecayMsg = "";
      if (this.weightDecay !== null) {
        const currentWd = this._calculateWeightDecay(this.currentStep - 1);
        weightDecayMsg = `, weight_decay: ${currentWd.toFixed(6)}`;
      }
      
      logQueue.put(`Epoch ${epoch+1} completed with lr: ${currentLr.toFixed(6)}${momentumMsg}${weightDecayMsg}`);
    }
    
    // Update UI
    if (this.onLrUpdate && model && model.optimizer) {
      try {
        this.onLrUpdate(model.optimizer.learningRate);
      } catch (e) {
        console.error("Error updating LR display:", e);
      }
    }
  }
  
  /**
   * Called at the end of training
   */
  onTrainEnd() {
    if (this.verbose > 0) {
      logQueue.put("OneCycleLR schedule completed");
    }
  }
  
  /**
   * Returns the full learning rate schedule as an array for plotting
   * @return {Array} Array of learning rates for each step
   */
  getLrSchedule() {
    return Array.from({length: this.totalSteps}, (_, i) => this._calculateLr(i));
  }
  
  /**
   * Returns the full momentum schedule as an array for plotting
   * @return {Array} Array of momentum values for each step
   */
  getMomentumSchedule() {
    return Array.from({length: this.totalSteps}, (_, i) => this._calculateMomentum(i));
  }
  
  /**
   * Returns the full weight decay schedule as an array for plotting
   * @return {Array|null} Array of weight decay values for each step, or null if weight decay is disabled
   */
  getWeightDecaySchedule() {
    if (this.weightDecay === null) {
      return null;
    }
    return Array.from({length: this.totalSteps}, (_, i) => this._calculateWeightDecay(i));
  }
}

// Export OneCycleLR class
export { OneCycleLR };